File: ntsc_simulator/decoder.py
"""NTSC composite video decoder: 1D composite signal -> RGB frames."""

import numpy as np
from scipy.signal import firwin
from scipy.fft import rfft, irfft, next_fast_len

from .constants import (
    FSC, SAMPLE_RATE, SAMPLES_PER_LINE, TOTAL_LINES, VISIBLE_LINES,
    FRONT_PORCH_SAMPLES, HSYNC_SAMPLES, BREEZEWAY_SAMPLES,
    BURST_SAMPLES, BACK_PORCH_SAMPLES, ACTIVE_SAMPLES,
    COMPOSITE_SCALE, COMPOSITE_OFFSET,
    YIQ_TO_RGB, I_PHASE_RAD, Q_PHASE_RAD,
    I_BW, Q_BW, LUMA_BW,
)

_F = np.float32
_OMEGA_PER_SAMPLE = _F(2.0 * np.pi * FSC / SAMPLE_RATE)

# Precompute filter coefficients (float32, 101 taps)
_NYQ = SAMPLE_RATE / 2
_NUM_TAPS = 101
_FIR_Y = firwin(_NUM_TAPS, LUMA_BW / _NYQ).astype(np.float32)
_FIR_I = firwin(_NUM_TAPS, I_BW / _NYQ).astype(np.float32)
_FIR_Q = firwin(_NUM_TAPS, Q_BW / _NYQ).astype(np.float32)

# Precompute offsets
_ACTIVE_START = (FRONT_PORCH_SAMPLES + HSYNC_SAMPLES +
                 BREEZEWAY_SAMPLES + BURST_SAMPLES + BACK_PORCH_SAMPLES)
_BURST_START = FRONT_PORCH_SAMPLES + HSYNC_SAMPLES + BREEZEWAY_SAMPLES
_ACTIVE_INDICES = np.arange(ACTIVE_SAMPLES, dtype=np.float32) + _ACTIVE_START
_BURST_INDICES = np.arange(BURST_SAMPLES, dtype=np.float32) + _BURST_START

_RIGHT_PAD = _NUM_TAPS

# Float32 constants for hot-path arithmetic
_COMP_SCALE = _F(COMPOSITE_SCALE)
_COMP_OFFSET = _F(COMPOSITE_OFFSET)
_I_PHASE = _F(I_PHASE_RAD)
_Q_PHASE = _F(Q_PHASE_RAD)
_PI = _F(np.pi)

# Precompute visible line -> absolute line mapping
_ABS_LINES = np.empty(VISIBLE_LINES, dtype=np.int32)
_ABS_LINES[0::2] = np.arange(20, 260)
_ABS_LINES[1::2] = np.arange(283, 523)

# Precompute 1H comb reference lines (adjacent line in same field).
# First line of each field uses the next line since there's no previous.
_REF_LINES = _ABS_LINES - 1
_REF_LINES[0] = _ABS_LINES[0] + 1   # field 1 first line: use next
_REF_LINES[1] = _ABS_LINES[1] + 1   # field 2 first line: use next


def _fft_filtfilt_2d(coeffs, data_2d):
    """FFT-based zero-phase FIR filter applied to each row of a 2D array.

    Equivalent to scipy.signal.filtfilt(coeffs, 1.0, data_2d, axis=1) but
    computes |H(f)|^2 * X(f) via FFT for O(N log N) instead of O(N*M).

    For a symmetric FIR filter, filtfilt's forward-backward application
    yields zero phase shift and squared magnitude response |H(f)|^2.
    """
    n_cols = data_2d.shape[1]
    fft_n = next_fast_len(n_cols + len(coeffs) - 1)
    H = rfft(coeffs, n=fft_n)
    H2 = (H * np.conj(H)).real  # |H(f)|^2 — real-valued
    X = rfft(data_2d, n=fft_n, axis=1)
    return irfft(X * H2, n=fft_n, axis=1)[:, :n_cols]


def decode_frame(signal, frame_number=0, output_width=640, output_height=480,
                 comb_1h=False):
    """Decode a composite NTSC signal back to an RGB frame (vectorized).

    Args:
        signal: 1D composite signal array for one frame (525 lines).
        frame_number: Frame index.
        output_width: Output frame width in pixels.
        output_height: Output frame height in pixels.
        comb_1h: Use 1H (one-line delay) comb filter instead of horizontal
                 2-sample comb. Reduces cross-color rainbow artifacts but
                 introduces hanging-dot artifacts at vertical transitions.

    Returns:
        RGB frame as numpy array (output_height x output_width x 3, uint8).
    """
    # Reshape signal into lines
    signal_2d = signal[:TOTAL_LINES * SAMPLES_PER_LINE].reshape(TOTAL_LINES, SAMPLES_PER_LINE)

    # Convert to float32 if needed (e.g. when loaded from float64 .npy file)
    if signal_2d.dtype != np.float32:
        signal_2d = signal_2d.astype(np.float32)

    # Extract FULL lines for all visible lines (910 samples each).
    # Processing the entire line gives the comb filter and FIR filters
    # valid signal data to settle on before reaching the active region,
    # eliminating edge artifacts on the left side of the picture.
    full_lines = signal_2d[_ABS_LINES].copy()  # (480, 910)

    # --- Comb filter for luma/chroma separation (vectorized) ---
    if comb_1h:
        # 1H comb: use adjacent line in same field as reference.
        # Subcarrier phase flips π between consecutive lines, so adding
        # cancels chroma and subtracting cancels luma.
        ref_lines = signal_2d[_REF_LINES]  # (480, 910)
        y_full = (full_lines + ref_lines) * _F(0.5)
        chroma_full = (full_lines - ref_lines) * _F(0.5)
    else:
        # Horizontal 2-sample delay comb at 4xfsc
        delayed = np.zeros_like(full_lines)
        delayed[:, 2:] = full_lines[:, :-2]
        y_full = (full_lines + delayed) * _F(0.5)
        chroma_full = (full_lines - delayed) * _F(0.5)

    # Lowpass luma at 4.2 MHz to remove residual subcarrier energy
    y_full = _fft_filtfilt_2d(_FIR_Y, y_full)

    # Undo composite voltage scaling on luma
    y_full = (y_full - _COMP_OFFSET) / _COMP_SCALE

    # --- Burst phase detection (vectorized) ---
    burst_all = full_lines[:, _BURST_START:_BURST_START + BURST_SAMPLES]
    line_phases = _PI * _ABS_LINES.astype(np.float32)  # (480,)

    # Reference carriers at burst positions for all lines
    burst_omega = _OMEGA_PER_SAMPLE * _BURST_INDICES.reshape(1, -1) + line_phases.reshape(-1, 1)
    cos_ref = np.cos(burst_omega)  # (480, burst_samples)
    sin_ref = np.sin(burst_omega)

    cos_corr = np.sum(burst_all * cos_ref, axis=1)  # (480,)
    sin_corr = np.sum(burst_all * sin_ref, axis=1)
    burst_phase = np.arctan2(sin_corr, cos_corr) - _PI  # (480,)

    # --- Chroma demodulation (vectorized) ---
    # Zero out chroma in the blanking region (before active start).
    # The burst was already used for phase detection above, so we null it
    # to prevent the colorburst from leaking through the FIR filter into
    # the left edge of the active picture.
    chroma_full[:, :_ACTIVE_START] = 0.0

    # Carrier phase for each line's sample range
    full_indices = np.arange(SAMPLES_PER_LINE, dtype=np.float32)
    full_omega = (_OMEGA_PER_SAMPLE * full_indices.reshape(1, -1)
                  + (line_phases + burst_phase).reshape(-1, 1))

    # Product detection on full lines (910 samples)
    i_raw = _F(2.0) * chroma_full * np.cos(full_omega + _I_PHASE)
    q_raw = _F(2.0) * chroma_full * np.cos(full_omega + _Q_PHASE)

    # Pad the demodulated baseband I/Q on the right with reflected values.
    # The active region ends at sample 910 (= SAMPLES_PER_LINE) with no
    # blanking after it. Reflect padding preserves signal continuity for
    # the filter's backward pass better than edge or zero padding.
    i_raw = np.pad(i_raw, ((0, 0), (0, _RIGHT_PAD)), mode='reflect')
    q_raw = np.pad(q_raw, ((0, 0), (0, _RIGHT_PAD)), mode='reflect')

    # Low-pass filter padded lines, then strip padding
    i_full = _fft_filtfilt_2d(_FIR_I, i_raw)[:, :SAMPLES_PER_LINE]
    q_full = _fft_filtfilt_2d(_FIR_Q, q_raw)[:, :SAMPLES_PER_LINE]

    # Crop to active region after filtering
    y_all = y_full[:, _ACTIVE_START:_ACTIVE_START + ACTIVE_SAMPLES]
    i_demod = i_full[:, _ACTIVE_START:_ACTIVE_START + ACTIVE_SAMPLES]
    q_demod = q_full[:, _ACTIVE_START:_ACTIVE_START + ACTIVE_SAMPLES]

    # Undo composite scaling on chroma
    i_demod /= _COMP_SCALE
    q_demod /= _COMP_SCALE

    # --- YIQ to RGB (vectorized) ---
    yiq = np.stack([y_all, i_demod, q_demod], axis=-1)  # (480, 754, 3)
    rgb = yiq @ YIQ_TO_RGB.T

    np.clip(rgb, 0.0, 1.0, out=rgb)

    # Convert to uint8
    output = np.clip(rgb * 255.0, 0, 255).astype(np.uint8)

    # Resize to output dimensions
    if output_width != ACTIVE_SAMPLES or output_height != VISIBLE_LINES:
        import cv2
        output = cv2.resize(output, (output_width, output_height),
                            interpolation=cv2.INTER_LINEAR)

    return output

File: rust/src/decoder.rs
//! NTSC composite video decoder: composite signal -> RGB frames.

use realfft::RealFftPlanner;

use crate::constants::*;
use crate::filters::{self, FilterKernel, FilterScratch};

const RIGHT_PAD: usize = NUM_TAPS;

/// Precomputed data for the decoder.
/// Each thread should have its own Decoder to avoid contention on scratch buffers.
pub struct Decoder {
    /// Filter for full-line luma (910 samples)
    luma_filter: FilterKernel,
    /// Filters for I/Q (910 + RIGHT_PAD samples)
    i_filter: FilterKernel,
    q_filter: FilterKernel,
    /// Absolute line numbers for visible lines
    abs_lines: [usize; VISIBLE_LINES],
    /// Reference lines for 1H comb
    ref_lines: [usize; VISIBLE_LINES],
    /// FFT scratch for luma (sized for SAMPLES_PER_LINE FFT)
    luma_scratch: FilterScratch,
    /// FFT scratch for I/Q (sized for SAMPLES_PER_LINE + RIGHT_PAD FFT)
    iq_scratch: FilterScratch,
    // --- Pre-allocated working buffers ---
    full_lines: Vec<f32>,
    y_full: Vec<f32>,
    chroma_full: Vec<f32>,
    burst_phases: Vec<f32>,
    i_raw: Vec<f32>,
    q_raw: Vec<f32>,
    output: Vec<u8>,
    /// Cached output dimensions for buffer invalidation
    cached_out_dims: (usize, usize),
}

impl Decoder {
    pub fn new() -> Self {
        let mut planner = RealFftPlanner::<f32>::new();

        let fir_y = filters::design_lowpass(LUMA_BW, NUM_TAPS);
        let fir_i = filters::design_lowpass(I_BW, NUM_TAPS);
        let fir_q = filters::design_lowpass(Q_BW, NUM_TAPS);

        let iq_padded_len = SAMPLES_PER_LINE + RIGHT_PAD;

        let luma_filter = FilterKernel::new(&fir_y, SAMPLES_PER_LINE);
        let i_filter = FilterKernel::new(&fir_i, iq_padded_len);

        let luma_fft_n = luma_filter.fft_n();
        let iq_fft_n = i_filter.fft_n();

        Self {
            luma_filter,
            i_filter,
            q_filter: FilterKernel::new(&fir_q, iq_padded_len),
            abs_lines: build_abs_lines(),
            ref_lines: build_ref_lines(),
            luma_scratch: FilterScratch::with_planner(&mut planner, luma_fft_n),
            iq_scratch: FilterScratch::with_planner(&mut planner, iq_fft_n),
            full_lines: vec![0.0f32; VISIBLE_LINES * SAMPLES_PER_LINE],
            y_full: vec![0.0f32; VISIBLE_LINES * SAMPLES_PER_LINE],
            chroma_full: vec![0.0f32; VISIBLE_LINES * SAMPLES_PER_LINE],
            burst_phases: vec![0.0f32; VISIBLE_LINES],
            i_raw: vec![0.0f32; VISIBLE_LINES * iq_padded_len],
            q_raw: vec![0.0f32; VISIBLE_LINES * iq_padded_len],
            output: Vec::new(),
            cached_out_dims: (0, 0),
        }
    }

    /// Decode a composite NTSC signal back to an RGB frame.
    /// Returns a reference to the internal output buffer (valid until the next call).
    pub fn decode_frame(
        &mut self,
        signal: &[f32],
        output_width: usize,
        output_height: usize,
        comb_1h: bool,
    ) -> &[u8] {
        let spl = SAMPLES_PER_LINE;

        // Ensure output buffer is the right size
        let out_size = output_width * output_height * 3;
        if self.cached_out_dims != (output_width, output_height) {
            self.output.resize(out_size, 0);
            self.cached_out_dims = (output_width, output_height);
        }

        // Extract full lines for all visible lines (480 x 910)
        for vis in 0..VISIBLE_LINES {
            let abs = self.abs_lines[vis];
            let sig_start = abs * spl;
            let dst_start = vis * spl;
            self.full_lines[dst_start..dst_start + spl]
                .copy_from_slice(&signal[sig_start..sig_start + spl]);
        }

        // Comb filter for luma/chroma separation
        if comb_1h {
            for vis in 0..VISIBLE_LINES {
                let ref_line = self.ref_lines[vis];
                let ref_start = ref_line * spl;
                let cur_start = vis * spl;
                for n in 0..spl {
                    let cur = self.full_lines[cur_start + n];
                    let ref_val = signal[ref_start + n];
                    self.y_full[cur_start + n] = (cur + ref_val) * 0.5;
                    self.chroma_full[cur_start + n] = (cur - ref_val) * 0.5;
                }
            }
        } else {
            for vis in 0..VISIBLE_LINES {
                let base = vis * spl;
                self.y_full[base] = self.full_lines[base] * 0.5;
                self.y_full[base + 1] = self.full_lines[base + 1] * 0.5;
                self.chroma_full[base] = self.full_lines[base] * 0.5;
                self.chroma_full[base + 1] = self.full_lines[base + 1] * 0.5;

                for n in 2..spl {
                    let cur = self.full_lines[base + n];
                    let delayed = self.full_lines[base + n - 2];
                    self.y_full[base + n] = (cur + delayed) * 0.5;
                    self.chroma_full[base + n] = (cur - delayed) * 0.5;
                }
            }
        }

        // Lowpass luma at 4.2 MHz
        filters::filter_rows_sequential(
            &self.luma_filter,
            &mut self.y_full,
            spl,
            &mut self.luma_scratch,
        );

        // Undo composite voltage scaling on luma
        for v in &mut self.y_full {
            *v = (*v - COMPOSITE_OFFSET) / COMPOSITE_SCALE;
        }

        // Burst phase detection using carrier LUT
        for vis in 0..VISIBLE_LINES {
            let abs_line = self.abs_lines[vis];
            let line_phase = std::f32::consts::PI * abs_line as f32;
            let base = vis * spl;

            let phi = std::f32::consts::FRAC_PI_2 * BURST_START as f32 + line_phase;
            let cos_phi = phi.cos();
            let sin_phi = phi.sin();

            let cos_lut = [cos_phi, -sin_phi, -cos_phi, sin_phi];
            let sin_lut = [sin_phi, cos_phi, -sin_phi, -cos_phi];

            let mut cos_corr = 0.0f32;
            let mut sin_corr = 0.0f32;

            for n in 0..BURST_SAMPLES {
                let burst_val = self.full_lines[base + BURST_START + n];
                cos_corr += burst_val * cos_lut[n & 3];
                sin_corr += burst_val * sin_lut[n & 3];
            }

            self.burst_phases[vis] = sin_corr.atan2(cos_corr) - std::f32::consts::PI;
        }

        // Zero out chroma in blanking region
        for vis in 0..VISIBLE_LINES {
            let base = vis * spl;
            self.chroma_full[base..base + ACTIVE_START].fill(0.0);
        }

        // Product detection using carrier LUT, then pad right with reflect
        let iq_padded_len = spl + RIGHT_PAD;

        for vis in 0..VISIBLE_LINES {
            let abs_line = self.abs_lines[vis];
            let line_phase = std::f32::consts::PI * abs_line as f32;
            let bp = self.burst_phases[vis];
            let chroma_base = vis * spl;
            let iq_base = vis * iq_padded_len;

            let i_phi = line_phase + bp + I_PHASE_RAD;
            let q_phi = line_phase + bp + Q_PHASE_RAD;

            let i_cos = i_phi.cos();
            let i_sin = i_phi.sin();
            let q_cos = q_phi.cos();
            let q_sin = q_phi.sin();

            let i_lut = [i_cos, -i_sin, -i_cos, i_sin];
            let q_lut = [q_cos, -q_sin, -q_cos, q_sin];

            for n in 0..spl {
                let chroma_val = self.chroma_full[chroma_base + n];
                self.i_raw[iq_base + n] = 2.0 * chroma_val * i_lut[n & 3];
                self.q_raw[iq_base + n] = 2.0 * chroma_val * q_lut[n & 3];
            }

            // Reflect padding on right
            for j in 0..RIGHT_PAD {
                self.i_raw[iq_base + spl + j] = self.i_raw[iq_base + spl - 2 - j];
                self.q_raw[iq_base + spl + j] = self.q_raw[iq_base + spl - 2 - j];
            }
        }

        // Low-pass filter I and Q
        filters::filter_rows_sequential(
            &self.i_filter,
            &mut self.i_raw,
            iq_padded_len,
            &mut self.iq_scratch,
        );
        filters::filter_rows_sequential(
            &self.q_filter,
            &mut self.q_raw,
            iq_padded_len,
            &mut self.iq_scratch,
        );

        // Crop to active region and convert YIQ -> RGB
        let need_resize = output_width != ACTIVE_SAMPLES || output_height != VISIBLE_LINES;

        if !need_resize {
            for vis in 0..VISIBLE_LINES {
                let y_base = vis * spl + ACTIVE_START;
                let iq_base = vis * iq_padded_len + ACTIVE_START;
                let out_base = vis * ACTIVE_SAMPLES * 3;

                for n in 0..ACTIVE_SAMPLES {
                    let y = self.y_full[y_base + n];
                    let i = self.i_raw[iq_base + n] / COMPOSITE_SCALE;
                    let q = self.q_raw[iq_base + n] / COMPOSITE_SCALE;

                    let r = (YIQ_TO_RGB[0][0] * y + YIQ_TO_RGB[0][1] * i + YIQ_TO_RGB[0][2] * q).clamp(0.0, 1.0);
                    let g = (YIQ_TO_RGB[1][0] * y + YIQ_TO_RGB[1][1] * i + YIQ_TO_RGB[1][2] * q).clamp(0.0, 1.0);
                    let b = (YIQ_TO_RGB[2][0] * y + YIQ_TO_RGB[2][1] * i + YIQ_TO_RGB[2][2] * q).clamp(0.0, 1.0);

                    self.output[out_base + n * 3] = (r * 255.0 + 0.5) as u8;
                    self.output[out_base + n * 3 + 1] = (g * 255.0 + 0.5) as u8;
                    self.output[out_base + n * 3 + 2] = (b * 255.0 + 0.5) as u8;
                }
            }
        } else {
            // Bilinear resize with fused YIQ->RGB
            let x_scale = if output_width > 1 {
                (ACTIVE_SAMPLES - 1) as f32 / (output_width - 1) as f32
            } else {
                0.0
            };
            let y_scale = if output_height > 1 {
                (VISIBLE_LINES - 1) as f32 / (output_height - 1) as f32
            } else {
                0.0
            };

            let inv_comp_scale = 1.0 / COMPOSITE_SCALE;

            for dy in 0..output_height {
                let sy = dy as f32 * y_scale;
                let sy0 = (sy as usize).min(VISIBLE_LINES - 1);
                let sy1 = (sy0 + 1).min(VISIBLE_LINES - 1);
                let yt = sy - sy0 as f32;
                let yt_inv = 1.0 - yt;

                for dx in 0..output_width {
                    let sx = dx as f32 * x_scale;
                    let sx0 = (sx as usize).min(ACTIVE_SAMPLES - 1);
                    let sx1 = (sx0 + 1).min(ACTIVE_SAMPLES - 1);
                    let xt = sx - sx0 as f32;
                    let xt_inv = 1.0 - xt;

                    let w00 = xt_inv * yt_inv;
                    let w01 = xt * yt_inv;
                    let w10 = xt_inv * yt;
                    let w11 = xt * yt;

                    let y = self.y_full[sy0 * spl + ACTIVE_START + sx0] * w00
                        + self.y_full[sy0 * spl + ACTIVE_START + sx1] * w01
                        + self.y_full[sy1 * spl + ACTIVE_START + sx0] * w10
                        + self.y_full[sy1 * spl + ACTIVE_START + sx1] * w11;

                    let i = (self.i_raw[sy0 * iq_padded_len + ACTIVE_START + sx0] * w00
                        + self.i_raw[sy0 * iq_padded_len + ACTIVE_START + sx1] * w01
                        + self.i_raw[sy1 * iq_padded_len + ACTIVE_START + sx0] * w10
                        + self.i_raw[sy1 * iq_padded_len + ACTIVE_START + sx1] * w11)
                        * inv_comp_scale;

                    let q = (self.q_raw[sy0 * iq_padded_len + ACTIVE_START + sx0] * w00
                        + self.q_raw[sy0 * iq_padded_len + ACTIVE_START + sx1] * w01
                        + self.q_raw[sy1 * iq_padded_len + ACTIVE_START + sx0] * w10
                        + self.q_raw[sy1 * iq_padded_len + ACTIVE_START + sx1] * w11)
                        * inv_comp_scale;

                    let r = (YIQ_TO_RGB[0][0] * y + YIQ_TO_RGB[0][1] * i + YIQ_TO_RGB[0][2] * q).clamp(0.0, 1.0);
                    let g = (YIQ_TO_RGB[1][0] * y + YIQ_TO_RGB[1][1] * i + YIQ_TO_RGB[1][2] * q).clamp(0.0, 1.0);
                    let b = (YIQ_TO_RGB[2][0] * y + YIQ_TO_RGB[2][1] * i + YIQ_TO_RGB[2][2] * q).clamp(0.0, 1.0);

                    let idx = (dy * output_width + dx) * 3;
                    self.output[idx] = (r * 255.0 + 0.5) as u8;
                    self.output[idx + 1] = (g * 255.0 + 0.5) as u8;
                    self.output[idx + 2] = (b * 255.0 + 0.5) as u8;
                }
            }
        }

        &self.output
    }
}

impl Default for Decoder {
    fn default() -> Self {
        Self::new()
    }
}
